Spdlog can't compile on alpine linux because it doesn't detect musl correctly

```
FAILED: src/logging/CMakeFiles/logging.dir/logger.cpp.o 
ccache /usr/bin/c++    -I/usr/local/include -I/project/src/. -Isrc -isystem /project/3rd-party/spdlog/include -Wall -Werror -Wshadow -Wextra   -std=gnu++14 -MD -MT src/logging/CMakeFiles/logging.dir/logger.cpp.o -MF src/logging/CMakeFiles/logging.dir/logger.cpp.o.d -o src/logging/CMakeFiles/logging.dir/logger.cpp.o -c /project/src/logging/logger.cpp
In file included from /project/3rd-party/spdlog/include/spdlog/details/log_msg.h:9:0,
                 from /project/3rd-party/spdlog/include/spdlog/sinks/sink.h:9,
                 from /project/3rd-party/spdlog/include/spdlog/sinks/base_sink.h:13,
                 from /project/3rd-party/spdlog/include/spdlog/logger.h:15,
                 from /project/3rd-party/spdlog/include/spdlog/spdlog.h:14,
                 from /project/src/logging/logger.h:5,
                 from /project/src/logging/logger.cpp:2:
/project/3rd-party/spdlog/include/spdlog/details/os.h: In function 'std::__cxx11::string spdlog::details::os::errno_str(int)':
/project/3rd-party/spdlog/include/spdlog/details/os.h:389:58: error: no matching function for call to 'std::__cxx11::basic_string<char>::basic_string(int)'
     return std::string(strerror_r(err_num, buf, buf_size));
                                                          ^
In file included from /usr/include/c++/6.2.1/string:52:0,
                 from /usr/include/c++/6.2.1/stdexcept:39,
                 from /usr/include/c++/6.2.1/array:39,
                 from /usr/include/c++/6.2.1/tuple:39,
                 from /usr/include/c++/6.2.1/functional:55,
                 from /usr/include/c++/6.2.1/memory:79,
                 from /project/3rd-party/spdlog/include/spdlog/fmt/bundled/format.h:37,
                 from /project/3rd-party/spdlog/include/spdlog/fmt/fmt.h:21,
                 from /project/3rd-party/spdlog/include/spdlog/fmt/ostr.h:11,
                 from /project/src/logging/logger.h:4,
                 from /project/src/logging/logger.cpp:2:
/usr/include/c++/6.2.1/bits/basic_string.h:549:9: note: candidate: template<class _InputIterator, class> std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::basic_string(_InputIterator, _InputIterator, const _Alloc&)
         basic_string(_InputIterator __beg, _InputIterator __end,
         ^~~~~~~~~~~~
/usr/include/c++/6.2.1/bits/basic_string.h:549:9: note:   template argument deduction/substitution failed:
In file included from /project/3rd-party/spdlog/include/spdlog/details/log_msg.h:9:0,
                 from /project/3rd-party/spdlog/include/spdlog/sinks/sink.h:9,
                 from /project/3rd-party/spdlog/include/spdlog/sinks/base_sink.h:13,
                 from /project/3rd-party/spdlog/include/spdlog/logger.h:15,
                 from /project/3rd-party/spdlog/include/spdlog/spdlog.h:14,
                 from /project/src/logging/logger.h:5,
                 from /project/src/logging/logger.cpp:2:
/project/3rd-party/spdlog/include/spdlog/details/os.h:389:58: note:   candidate expects 3 arguments, 1 provided
     return std::string(strerror_r(err_num, buf, buf_size));
                                                          ^
In file included from /usr/include/c++/6.2.1/string:52:0,
                 from /usr/include/c++/6.2.1/stdexcept:39,
                 from /usr/include/c++/6.2.1/array:39,
                 from /usr/include/c++/6.2.1/tuple:39,
                 from /usr/include/c++/6.2.1/functional:55,
                 from /usr/include/c++/6.2.1/memory:79,
                 from /project/3rd-party/spdlog/include/spdlog/fmt/bundled/format.h:37,
                 from /project/3rd-party/spdlog/include/spdlog/fmt/fmt.h:21,
                 from /project/3rd-party/spdlog/include/spdlog/fmt/ostr.h:11,
                 from /project/src/logging/logger.h:4,
                 from /project/src/logging/logger.cpp:2:
/usr/include/c++/6.2.1/bits/basic_string.h:511:7: note: candidate: std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::basic_string(std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&&, const _Alloc&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]
       basic_string(basic_string&& __str, const _Alloc& __a)
       ^~~~~~~~~~~~
/usr/include/c++/6.2.1/bits/basic_string.h:511:7: note:   candidate expects 2 arguments, 1 provided
/usr/include/c++/6.2.1/bits/basic_string.h:507:7: note: candidate: std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::basic_string(const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&, const _Alloc&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]
       basic_string(const basic_string& __str, const _Alloc& __a)
       ^~~~~~~~~~~~
/usr/include/c++/6.2.1/bits/basic_string.h:507:7: note:   candidate expects 2 arguments, 1 provided
/usr/include/c++/6.2.1/bits/basic_string.h:503:7: note: candidate: std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::basic_string(std::initializer_list<_Tp>, const _Alloc&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]
       basic_string(initializer_list<_CharT> __l, const _Alloc& __a = _Alloc())
       ^~~~~~~~~~~~
/usr/include/c++/6.2.1/bits/basic_string.h:503:7: note:   no known conversion for argument 1 from 'int' to 'std::initializer_list<char>'
/usr/include/c++/6.2.1/bits/basic_string.h:476:7: note: candidate: std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::basic_string(std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]
       basic_string(basic_string&& __str) noexcept
       ^~~~~~~~~~~~
/usr/include/c++/6.2.1/bits/basic_string.h:476:7: note:   no known conversion for argument 1 from 'int' to 'std::__cxx11::basic_string<char>&&'
/usr/include/c++/6.2.1/bits/basic_string.h:464:7: note: candidate: std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::basic_string(std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::size_type, _CharT, const _Alloc&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]
       basic_string(size_type __n, _CharT __c, const _Alloc& __a = _Alloc())
       ^~~~~~~~~~~~
/usr/include/c++/6.2.1/bits/basic_string.h:464:7: note:   candidate expects 3 arguments, 1 provided
/usr/include/c++/6.2.1/bits/basic_string.h:454:7: note: candidate: std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::basic_string(const _CharT*, const _Alloc&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>] <near match>
       basic_string(const _CharT* __s, const _Alloc& __a = _Alloc())
       ^~~~~~~~~~~~
/usr/include/c++/6.2.1/bits/basic_string.h:454:7: note:   conversion of argument 1 would be ill-formed:
In file included from /project/3rd-party/spdlog/include/spdlog/details/log_msg.h:9:0,
                 from /project/3rd-party/spdlog/include/spdlog/sinks/sink.h:9,
                 from /project/3rd-party/spdlog/include/spdlog/sinks/base_sink.h:13,
                 from /project/3rd-party/spdlog/include/spdlog/logger.h:15,
                 from /project/3rd-party/spdlog/include/spdlog/spdlog.h:14,
                 from /project/src/logging/logger.h:5,
                 from /project/src/logging/logger.cpp:2:
/project/3rd-party/spdlog/include/spdlog/details/os.h:389:34: error: invalid conversion from 'int' to 'const char*' [-fpermissive]
     return std::string(strerror_r(err_num, buf, buf_size));
                        ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6.2.1/string:52:0,
                 from /usr/include/c++/6.2.1/stdexcept:39,
                 from /usr/include/c++/6.2.1/array:39,
                 from /usr/include/c++/6.2.1/tuple:39,
                 from /usr/include/c++/6.2.1/functional:55,
                 from /usr/include/c++/6.2.1/memory:79,
                 from /project/3rd-party/spdlog/include/spdlog/fmt/bundled/format.h:37,
                 from /project/3rd-party/spdlog/include/spdlog/fmt/fmt.h:21,
                 from /project/3rd-party/spdlog/include/spdlog/fmt/ostr.h:11,
                 from /project/src/logging/logger.h:4,
                 from /project/src/logging/logger.cpp:2:
/usr/include/c++/6.2.1/bits/basic_string.h:444:7: note: candidate: std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::basic_string(const _CharT*, std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::size_type, const _Alloc&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]
       basic_string(const _CharT* __s, size_type __n,
       ^~~~~~~~~~~~
/usr/include/c++/6.2.1/bits/basic_string.h:444:7: note:   candidate expects 3 arguments, 1 provided
/usr/include/c++/6.2.1/bits/basic_string.h:426:7: note: candidate: std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::basic_string(const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&, std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::size_type, std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::size_type, const _Alloc&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]
       basic_string(const basic_string& __str, size_type __pos,
       ^~~~~~~~~~~~
/usr/include/c++/6.2.1/bits/basic_string.h:426:7: note:   candidate expects 4 arguments, 1 provided
/usr/include/c++/6.2.1/bits/basic_string.h:410:7: note: candidate: std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::basic_string(const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&, std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::size_type, std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::size_type) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]
       basic_string(const basic_string& __str, size_type __pos,
       ^~~~~~~~~~~~
/usr/include/c++/6.2.1/bits/basic_string.h:410:7: note:   candidate expects 3 arguments, 1 provided
/usr/include/c++/6.2.1/bits/basic_string.h:397:7: note: candidate: std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::basic_string(const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]
       basic_string(const basic_string& __str)
       ^~~~~~~~~~~~
/usr/include/c++/6.2.1/bits/basic_string.h:397:7: note:   no known conversion for argument 1 from 'int' to 'const std::__cxx11::basic_string<char>&'
/usr/include/c++/6.2.1/bits/basic_string.h:389:7: note: candidate: std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::basic_string(const _Alloc&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]
       basic_string(const _Alloc& __a) _GLIBCXX_NOEXCEPT
       ^~~~~~~~~~~~
/usr/include/c++/6.2.1/bits/basic_string.h:389:7: note:   no known conversion for argument 1 from 'int' to 'const std::allocator<char>&'
/usr/include/c++/6.2.1/bits/basic_string.h:380:7: note: candidate: std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::basic_string() [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]
       basic_string()
       ^~~~~~~~~~~~
/usr/include/c++/6.2.1/bits/basic_string.h:380:7: note:   candidate expects 0 arguments, 1 provided
ninja: build stopped: subcommand failed.
```

--- a/include/spdlog/details/os.h
+++ b/include/spdlog/details/os.h
@@ -364,6 +364,17 @@ inline std::string filename_to_str(const filename_t& filename)
 }
 #endif
 
+inline std::string errno_to_string(char [256], char* res) {
+    return std::string(res);
+}
+    
+inline std::string errno_to_string(char buf[256], int res) {
+    if (res == 0) {
+        return std::string(buf);
+    } else {
+        return "Unknown error";
+    }
+}
 
 // Return errno string (thread safe)
 inline std::string errno_str(int err_num)
@@ -386,7 +397,8 @@ inline std::string errno_str(int err_num)
         return "Unknown error";
 
 #else  // gnu version (might not use the given buf, so its retval pointer must be used)
-    return std::string(strerror_r(err_num, buf, buf_size));
+    auto err = strerror_r(err_num, buf, buf_size); // let compiler choose type
+    return errno_to_string(buf, err); // use overloading to select correct stringify function
 #endif
 }
 

